// typescript-eslint-plugin/lib/Diagnostics/LintDiagnosticMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LintDiagnosticMap = void 0;
/**
 * Provides a set of lint-diagnostics.
 */
class LintDiagnosticMap {
    constructor() {
        /**
         * The problems.
         */
        this.map = new Map();
    }
    /**
     * Gets the problem with the specified start and end.
     *
     * @param start
     * The start-position of the problem.
     *
     * @param end
     * The end-position of the problem.
     *
     * @returns
     * The diagnostic at the specified range.
     */
    Get(start, end) {
        return this.map.get(this.Key(start, end));
    }
    /**
     * Adds a problem to the map with the specified `start` and `end`.
     *
     * @param start
     * The start-position of the problem.
     *
     * @param end
     * The end-position of the problem.
     *
     * @param item
     * The item to add.
     */
    Set(start, end, item) {
        this.map.set(this.Key(start, end), item);
    }
    /**
     * Gets the problems.
     */
    get Values() {
        return this.map.values();
    }
    /**
     * Determines a key for a problem.
     *
     * @param start
     * The start-position of the problem.
     *
     * @param end
     * The end-position of the problem.
     *
     * @returns
     * The key for the specified range.
     */
    Key(start, end) {
        return JSON.stringify([start, end]);
    }
}
exports.LintDiagnosticMap = LintDiagnosticMap;
